File: src/main/java/com/indico/mutation/WorkflowSubmission.java
package com.indico.mutation;

import com.apollographql.apollo.ApolloCall;
import com.apollographql.apollo.api.Error;
import com.apollographql.apollo.api.Response;
import com.indico.Async;
import com.indico.IndicoClient;
import com.indico.Mutation;
import com.indico.RetryInterceptor;
import com.indico.WorkflowSubmissionGraphQLMutation;
import com.indico.storage.UploadFile;
import com.indico.type.FileInput;


import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;



public class WorkflowSubmission implements Mutation<List<Integer>> {

    private final IndicoClient client;
    private List<String> files;
    private int id;
    private final Logger logger = LogManager.getLogger(WorkflowSubmission.class);

    public WorkflowSubmission(IndicoClient client) {
        this.client = client;
    }
    
    /**
     * List of local file paths to submit
     * @param files File paths
     * @return WorkflowSubmission
     */
    public WorkflowSubmission files(List<String> files) {
        this.files = files;
        return this;
    }

    /**
     * Id of workflow to submit files to
     * @param id Workflow Id
     * @return WorkflowSubmission
     */
    public WorkflowSubmission workflowId(int id) {
        this.id = id;
        return this;
    }

    /**
     * Executes request and returns Submissions
     * @return Integer List
     */
    @Override
    public List<Integer> execute() {
        JSONArray fileMetadata;
        List<FileInput> files = new ArrayList<FileInput>();
        try {
            fileMetadata = this.upload(this.files);
            for (Object f : fileMetadata) {
                JSONObject uploadMeta = (JSONObject) f;
                JSONObject meta = new JSONObject();
                meta.put("name", uploadMeta.getString("name"));
                meta.put("path", uploadMeta.getString("path"));
                meta.put("upload_type", uploadMeta.getString("upload_type"));
                FileInput input = FileInput.builder().filename(((JSONObject) f).getString("name")).filemeta(meta).build();
                files.add(input);
            }
        } catch (IOException e) {
            throw new RuntimeException(e.getMessage(), e.fillInStackTrace());
        }

        ApolloCall<WorkflowSubmissionGraphQLMutation.Data> apolloCall = this.client.apolloClient.mutate(WorkflowSubmissionGraphQLMutation.builder()
                .files(files)
                .workflowId(this.id)
                .build());

        Response<WorkflowSubmissionGraphQLMutation.Data> response = (Response<WorkflowSubmissionGraphQLMutation.Data>) Async.executeSync(apolloCall, this.client.config).join();
        if (response.hasErrors()) {
            StringBuilder errors = new StringBuilder();
            for (Error err : response.errors()) {
                errors.append(err.toString() + "\n");
            }

            String msg = errors.toString();
            logger.error("Errors encountered extracting documents: " + msg);
            throw new RuntimeException("Failed to extract documents due to following error: \n" + msg);
        }
        WorkflowSubmissionGraphQLMutation.WorkflowSubmission workflowSubmission = response.data().workflowSubmission();
        return workflowSubmission.submissionIds();
    }

    private JSONArray upload(List<String> filePaths) throws IOException {
        UploadFile uploadRequest = new UploadFile(this.client);
        return uploadRequest.filePaths(filePaths).call();
    }
}

File: src/main/java/com/indico/mutation/WorkflowSubmission.kt
package com.indico.mutation

import com.expediagroup.graphql.client.jackson.types.OptionalInput
import com.indico.IndicoClient
import com.indico.exceptions.IndicoMutationException
import com.indico.graphql.WorkflowSubmissionGraphQL
import com.indico.graphql.inputs.FileInput
import com.indico.graphql.enums.SubmissionResultVersion as GraphQlResultVersion
import com.indico.type.SubmissionResultVersion
import org.apache.logging.log4j.LogManager
import java.io.IOException
import java.util.*

class WorkflowSubmission(private val client: IndicoClient) : Mutation<List<Int?>?, WorkflowSubmissionGraphQL.Result>() {
    private var files: List<String>? = null
    private var id = 0
    private var duplicationId: UUID? = null
    private var streams: Map<String, ByteArray>? = null
    private var resultFileVersion: SubmissionResultVersion? = null
    private val logger = LogManager.getLogger(javaClass)

    /**
     * List of local file paths to submit
     * @param files File paths
     * @return WorkflowSubmission
     */
    fun files(files: List<String>?): WorkflowSubmission {
        this.files = files
        return this
    }

    /**
     *
     * @param streams Dictionary of String identifiers and byte streams to upload.
     * @return
     */
    fun byteStreams(streams: Map<String, ByteArray>?): WorkflowSubmission {
        this.streams = streams
        return this
    }

    /**
     * Id of workflow to submit files to
     * @param id Workflow Id
     * @return WorkflowSubmission
     */
    fun workflowId(id: Int): WorkflowSubmission {
        this.id = id
        return this
    }

    /**
     * A UUID representing a unique set of files and workflow activity.
     * This optional parameter helps the platform detect and prevent duplicate submissions.
     * @param id
     * @return
     */
    fun duplicationId(id: UUID?): WorkflowSubmission {
        duplicationId = id
        return this
    }

    fun resultFileVersion(version: SubmissionResultVersion?): WorkflowSubmission{
        resultFileVersion = version
        return this
    }

    /**
     * Executes request and returns Submissions
     * @return Integer List
     */
    override fun execute(): List<Int?>? {
        val files: MutableList<FileInput> = ArrayList<FileInput>()
        try {
            if (this.files != null) {
                files.addAll(processFiles(this.files!!, client))
            }
            if (streams != null) {
                files.addAll(processStreams(this.streams!!, client))
            }
        } catch (e: IOException) {
            throw IndicoMutationException(e.message, e)
        }
        if (duplicationId == null) {
            duplicationId = UUID.randomUUID()
        }

        val resultVersion =
            if(resultFileVersion != null)
                OptionalInput.Defined(GraphQlResultVersion.valueOf(resultFileVersion.toString()))
            else OptionalInput.Undefined

        return try {
            val call = WorkflowSubmissionGraphQL(
                WorkflowSubmissionGraphQL.Variables(
                    files = files, workflowId = id, duplicationId = OptionalInput.Defined(duplicationId.toString()),
                    resultVersion = resultVersion
               ))
            val response = client.execute(call)
            handleErrors(response)
            val workflowSubmission = response.data!!.workflowSubmission!!
            if (workflowSubmission.isDuplicateRequest!!) {
                logger.debug(
                    "Duplicate submission sent for submission ids " + workflowSubmission.submissionIds.toString()
                )
            }
            workflowSubmission.submissionIds
        }catch (ex: java.lang.RuntimeException) {
            throw IndicoMutationException("Call to submit submission failed", ex)
        }
    }
}
